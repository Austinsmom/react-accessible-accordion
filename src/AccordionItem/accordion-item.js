// @flow

import React, { Component } from 'react';
import type { Node } from 'react';
import { inject, observer, Provider } from 'mobx-react';
import consecutive from 'consecutive';
import classNames from 'classnames';
import type { Store } from '../accordionStore/accordionStore';

const nextUuid = consecutive();

type AccordionItemProps = {
    children: Node,
    className: string,
    hideBodyClassName: string,
    itemkey: string | number,
    accordionStore: Store,
    disabled: boolean,
    expanded: boolean,
};

class AccordionItem extends Component<AccordionItemProps, *> {
    static defaultProps = {
        className: 'accordion__item',
        hideBodyClassName: '',
        disabled: false,
        expanded: false,
    };

    customKey = this.props.itemkey || nextUuid();

    componentWillMount() {
        this.props.accordionStore.addItem({
            itemkey: this.customKey,
            itemUuid: nextUuid(),
            expanded: this.props.expanded,
            disabled: this.props.disabled,
        });
    }

    componentWillUnmount() {
        this.props.accordionStore.removeItem(this.customKey);
    }

    // This is here so that the user can dynamically set the 'expanded' state using the 'expanded' prop.
    componentWillReceiveProps({
        expanded,
        accordionStore,
    }: AccordionItemProps) {
        if (expanded !== this.props.expanded) {
            accordionStore.setExpanded(this.customKey, expanded);
        }
    }

    render() {
        const {
            className,
            hideBodyClassName,
            children,
            accordionStore,
        } = this.props;

        const currentItem = accordionStore.items.find(
            item => item.itemkey === this.customKey,
        );

        if (!currentItem) {
            return null;
        }
        const { expanded } = currentItem;

        return (
            <Provider itemkey={this.customKey}>
                <div
                    className={classNames(className, {
                        [hideBodyClassName]: !expanded && hideBodyClassName,
                    })}
                >
                    {children}
                </div>
            </Provider>
        );
    }
}

export default inject('accordionStore')(observer(AccordionItem));
