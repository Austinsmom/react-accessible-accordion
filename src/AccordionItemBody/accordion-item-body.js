import React, { PropTypes } from 'react';
import classNames from 'classnames';

const defaultProps = {
    id: '',
    expanded: false,
    className: 'accordion__body',
    prefixCss: 'accordion__body',
    role: '',
};

const propTypes = {
    id: PropTypes.string,
    expanded: PropTypes.bool,
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.arrayOf(PropTypes.element),
    ]).isRequired,
    className: PropTypes.string,
    prefixCss: PropTypes.string,
    role: PropTypes.string,
};

const AccordionItemBody = ({ id, expanded, children, className, prefixCss, role }) => {
    const bodyClass = classNames(
        className,
        {
            [`${prefixCss}--hidden`]: !expanded,
        },
    );
    const ariaHidden = !expanded;
    return (
        <div
            id={id}
            className={bodyClass}
            aria-hidden={ariaHidden}
            role={role}
        >
            {children}
        </div>
    );
};

AccordionItemBody.propTypes = propTypes;
AccordionItemBody.defaultProps = defaultProps;

export default AccordionItemBody;
