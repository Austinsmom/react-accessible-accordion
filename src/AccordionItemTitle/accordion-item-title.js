// @flow

import React, { Component } from 'react';
import type { Node } from 'react';
import classNames from 'classnames';
import { inject, observer } from 'mobx-react';

type AccordionItemTitleProps = {
    children: Node,
    className: string,
    hideBodyClassName: string,
    accordionStore: {
        items: Array<Object>,
        accordion: boolean,
        onChange: Function,
        setExpanded: (string | number, boolean) => void,
    },
    itemkey: string | number,
};

type AccordionItemTitleState = {};

export class AccordionItemTitle extends Component<
    AccordionItemTitleProps,
    AccordionItemTitleState,
> {
    static accordionElementName = 'AccordionItemTitle';

    static defaultProps = {
        className: 'accordion__title',
        hideBodyClassName: '',
    };

    handleClick = () => {
        const { itemkey, accordionStore } = this.props;
        const { accordion, onChange, items } = accordionStore;
        const foundItem = items.find(item => item.itemkey === itemkey);
        if (!foundItem) return;

        if (accordion) {
            const newValue = Boolean(!foundItem.expanded);
            items.forEach(item => {
                this.props.accordionStore.setExpanded(item.itemkey, false);
                if (item.itemkey === foundItem.itemkey) {
                    this.props.accordionStore.setExpanded(
                        foundItem.itemkey,
                        newValue,
                    );
                }
            });

            onChange(foundItem.itemkey);
        } else {
            this.props.accordionStore.setExpanded(
                foundItem.itemkey,
                !foundItem.expanded,
            );
            onChange(
                this.props.accordionStore.items
                    .filter(item => item.expanded)
                    .map(item => item.itemkey),
            );
        }
    };

    handleKeyPress = (evt: SyntheticInputEvent<HTMLButtonElement>) => {
        if (evt.charCode === 13 || evt.charCode === 32) {
            this.handleClick();
        }
    };

    render() {
        const { items, accordion } = this.props.accordionStore;
        const { itemkey, children, className, hideBodyClassName } = this.props;
        const foundItem = items.find(item => item.itemkey === itemkey);
        if (!foundItem) return null;

        const { itemUuid, expanded } = foundItem;

        const id = `accordion__title-${itemUuid}`;
        const ariaControls = `accordion__body-${itemUuid}`;
        const role = accordion ? 'tab' : 'button';
        const titleClassName = classNames(className, {
            [hideBodyClassName]: hideBodyClassName && !expanded,
        });

        if (role === 'tab') {
            return (
                <div
                    id={id}
                    aria-selected={expanded}
                    aria-controls={ariaControls}
                    className={titleClassName}
                    onClick={this.handleClick}
                    role={role}
                    tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
                    onKeyPress={this.handleKeyPress}
                >
                    {children}
                </div>
            );
        }
        return (
            <div
                id={id}
                aria-expanded={expanded}
                aria-controls={ariaControls}
                className={titleClassName}
                onClick={this.handleClick}
                role={role}
                tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
                onKeyPress={this.handleKeyPress}
            >
                {children}
            </div>
        );
    }
}

export default inject('accordionStore', 'itemkey')(
    observer(AccordionItemTitle),
);
