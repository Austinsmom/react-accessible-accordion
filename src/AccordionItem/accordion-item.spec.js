// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import AccordionItemTitle from '../AccordionItemTitle/accordion-item-title';
import AccordionItemBody from '../AccordionItemBody/accordion-item-body';
import AccordionItem from './accordion-item';

jest.mock('uuid', () => {
    return {
        v4: () => 'mock-uuid-hash',
    };
});

jest.mock('../AccordionItemTitle/accordion-item-title', () => {
    const RealReact = require.requireActual('react');
    class MockElement extends RealReact.Component {
        render() {
            return RealReact.createElement('div', this.props, this.props.children);
        }
    }
    MockElement.accordionElementName = 'AccordionItemTitle';

    return MockElement;
});

jest.mock('../AccordionItemBody/accordion-item-body', () => {
    const RealReact = require.requireActual('react');
    class MockElement extends RealReact.Component {
        render() {
            return RealReact.createElement('div', this.props, this.props.children);
        }
    }
    MockElement.accordionElementName = 'AccordionItemBody';

    return MockElement;
});

describe('AccordionItem', () => {
    it('renders correctly with accordion true', () => {
        const tree = renderer.create(
            <AccordionItem>
                <AccordionItemTitle>
                    <div>Fake title</div>
                </AccordionItemTitle>
                <AccordionItemBody>
                    <div>Fake body</div>
                </AccordionItemBody>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with accordion false', () => {
        const tree = renderer.create(
            <AccordionItem accordion={false}>
                <AccordionItemTitle>
                    <div>Fake title</div>
                </AccordionItemTitle>
                <AccordionItemBody>
                    <div>Fake body</div>
                </AccordionItemBody>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('still renders with no AccordionItemTitle or AccordionItemBody', () => {
        const tree = renderer.create(
            <AccordionItem accordion={false}>
                <div>Fake title</div>
                <div>Fake body</div>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders with different className', () => {
        const tree = renderer.create(
            <AccordionItem className="testCSSClass">
                <AccordionItemTitle>
                    <div>Fake title</div>
                </AccordionItemTitle>
                <AccordionItemBody>
                    <div>Fake body</div>
                </AccordionItemBody>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders with different hideBodyClassName', () => {
        const tree = renderer.create(
            <AccordionItem expanded={false} hideBodyClassName="testCSSClass--hidden">
                <AccordionItemTitle>
                    <div>Fake title</div>
                </AccordionItemTitle>
                <AccordionItemBody>
                    <div>Fake body</div>
                </AccordionItemBody>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with other blocks inside', () => {
        const tree = renderer.create(
            <AccordionItem accordion={false}>
                <AccordionItemTitle>
                    <div>Fake title</div>
                </AccordionItemTitle>
                <div>Just another block</div>
                <AccordionItemBody>
                    <div>Fake body</div>
                </AccordionItemBody>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with other blocks inside 2', () => {
        const tree = renderer.create(
            <AccordionItem accordion={false}>
                <AccordionItemTitle>
                    <div>Fake title</div>
                </AccordionItemTitle>
                <AccordionItemBody>
                    <div>Fake body</div>
                </AccordionItemBody>
                <div>Just another block</div>
            </AccordionItem>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
