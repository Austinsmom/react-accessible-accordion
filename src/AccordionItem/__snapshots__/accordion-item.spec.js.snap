// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem renders correctly with accordion false 1`] = `
<div
  className="accordion__item"
>
  <div
    aria-controls="accordion__body-mock-uuid-hash"
    aria-expanded={false}
    className="accordion__title"
    id="accordion__title-mock-uuid-hash"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div>
      Fake title
    </div>
  </div>
  <div
    aria-hidden={true}
    aria-labelledby="accordion__title-mock-uuid-hash"
    className="accordion__body accordion__body--hidden"
    id="accordion__body-mock-uuid-hash"
    role={null}
  >
    <div>
      Fake body
    </div>
  </div>
</div>
`;

exports[`AccordionItem renders correctly with accordion true 1`] = `
<div
  className="accordion__item"
>
  <div
    aria-controls="accordion__body-mock-uuid-hash"
    aria-selected={false}
    className="accordion__title"
    id="accordion__title-mock-uuid-hash"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="tab"
    tabIndex="0"
  >
    <div>
      Fake title
    </div>
  </div>
  <div
    aria-hidden={true}
    aria-labelledby="accordion__title-mock-uuid-hash"
    className="accordion__body accordion__body--hidden"
    id="accordion__body-mock-uuid-hash"
    role="tabpanel"
  >
    <div>
      Fake body
    </div>
  </div>
</div>
`;

exports[`AccordionItem renders correctly with other blocks inside 1`] = `
<div
  className="accordion__item"
>
  <div
    aria-controls="accordion__body-mock-uuid-hash"
    aria-expanded={false}
    className="accordion__title"
    id="accordion__title-mock-uuid-hash"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div>
      Fake title
    </div>
  </div>
  <div>
    Just another block
  </div>
  <div
    aria-hidden={true}
    aria-labelledby="accordion__title-mock-uuid-hash"
    className="accordion__body accordion__body--hidden"
    id="accordion__body-mock-uuid-hash"
    role={null}
  >
    <div>
      Fake body
    </div>
  </div>
</div>
`;

exports[`AccordionItem renders correctly with other blocks inside 2 1`] = `
<div
  className="accordion__item"
>
  <div
    aria-controls="accordion__body-mock-uuid-hash"
    aria-expanded={false}
    className="accordion__title"
    id="accordion__title-mock-uuid-hash"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div>
      Fake title
    </div>
  </div>
  <div
    aria-hidden={true}
    aria-labelledby="accordion__title-mock-uuid-hash"
    className="accordion__body accordion__body--hidden"
    id="accordion__body-mock-uuid-hash"
    role={null}
  >
    <div>
      Fake body
    </div>
  </div>
  <div>
    Just another block
  </div>
</div>
`;

exports[`AccordionItem still renders with no AccordionItemTitle or AccordionItemBody 1`] = `
<div
  className="accordion__item"
>
  <div>
    Fake title
  </div>
  <div>
    Fake body
  </div>
</div>
`;

exports[`AccordionItem still renders with no children at all 1`] = `
<div
  className="accordion__item"
/>
`;
